### DISCUSSION TITLE
The time it takes to build...

### DISCUSSION THREAD

**Discussion by kmpoppe** [streetcomplete/StreetComplete#3376] [Q&A] [answered]
Hi app-devs! I just installed the newest Android Studio on my workplace notebook, Intel Core i9 with 64 gig of memory and only M2 flash drives that runs circles around most of the other developer's machines here, and yet the very first Grade build of the Dev App took 12 minutes none the less...

Is this supposed to be by design? Am I missing an important multi thread setting in the Studio?

Input's welcome!

Kai

**Comment by matkoniecz** [comment]
The first build will be extra-long as it fetches various dependencies. Subsequent should be faster.

But yes, build times are looong. I probably should look into it again, it is one of more irritating parts in development.

**Reply by kmpoppe** [reply]
The time I have waited for compilers to make up their minds in my life up until this point has reached months, so I'm not the impatient one but it feels really sluggish. I always wonder how development that costs and should earn actual money can have a successful business model that way ;-)

**Comment by smichel17** [comment]
I read on the Kotlin Slack channel this morning that in some cases, Kotlin build scripts can be slower than Groovy. People there said the difference was greater in situations that ~~don't apply to us~~ (many modules & using buildSrc), but it might be worth checking how big the difference actually is at some point.

edit: I take it back, we are actually using buildSrc. So maybe this makes a difference.

**Reply by westnordost** [reply]
> edit: I take it back, we are actually using buildSrc. So maybe this makes a difference.

This is in a sense also slow, because AFAIK the compilation of `buildSrc` will be done **every time** before ~the compilation of the app proper~ *executing any gradle task*. Maybe you can test: How long does a clean build take just like that and how long does it take if you delete the buildSrc folder? How long do subsequent builds take for these two cases?

**Comment by westnordost** [comment]
My machine (Intel i7-7700, 16 GB RAM, built on HDD):

A clean `gradle assembleDebug` took 3 minutes right now.
<details>
<summary>In Detail...</summary>
<table>
<thead>
<tr>
<th>Task</th>
<th>Duration</th>
<th>Result</th>
</tr>
</thead>
<tbody><tr>
<td>:app</td>
<td>3m32.31s</td>
<td>(total)</td>
</tr>
<tr>
<td>:app:compileDebugKotlin</td>
<td>1m1.96s</td>
<td></td>
</tr>
<tr>
<td>:app:kaptGenerateStubsDebugKotlin</td>
<td>41.784s</td>
<td></td>
</tr>
<tr>
<td>:app:minifyDebugWithR8</td>
<td>32.538s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugResources</td>
<td>30.804s</td>
<td></td>
</tr>
<tr>
<td>:app:l8DexDesugarLibDebug</td>
<td>10.810s</td>
<td></td>
</tr>
<tr>
<td>:app:kaptDebugKotlin</td>
<td>7.971s</td>
<td></td>
</tr>
<tr>
<td>:app:shrinkDebugRes</td>
<td>6.870s</td>
<td></td>
</tr>
<tr>
<td>:app:processDebugManifestForPackage</td>
<td>3.768s</td>
<td></td>
</tr>
<tr>
<td>:app:processDebugResources</td>
<td>3.033s</td>
<td></td>
</tr>
<tr>
<td>:app:compileDebugJavaWithJavac</td>
<td>2.905s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugAssets</td>
<td>1.723s</td>
<td></td>
</tr>
<tr>
<td>:app:compressDebugAssets</td>
<td>1.674s</td>
<td></td>
</tr>
<tr>
<td>:app:packageDebug</td>
<td>1.415s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugNativeLibs</td>
<td>1.301s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugJavaResource</td>
<td>0.964s</td>
<td></td>
</tr>
<tr>
<td>:app:checkDebugDuplicateClasses</td>
<td>0.877s</td>
<td></td>
</tr>
<tr>
<td>:app:dataBindingGenBaseClassesDebug</td>
<td>0.872s</td>
<td></td>
</tr>
<tr>
<td>:app:processDebugMainManifest</td>
<td>0.324s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugGeneratedProguardFiles</td>
<td>0.306s</td>
<td></td>
</tr>
<tr>
<td>:app:stripDebugDebugSymbols</td>
<td>0.115s</td>
<td></td>
</tr>
<tr>
<td>:app:dataBindingMergeDependencyArtifactsDebug</td>
<td>0.073s</td>
<td></td>
</tr>
<tr>
<td>:app:processDebugManifest</td>
<td>0.047s</td>
<td></td>
</tr>
<tr>
<td>:app:checkDebugAarMetadata</td>
<td>0.044s</td>
<td></td>
</tr>
<tr>
<td>:app:validateSigningDebug</td>
<td>0.041s</td>
<td></td>
</tr>
<tr>
<td>:app:generateDebugBuildConfig</td>
<td>0.016s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugShaders</td>
<td>0.014s</td>
<td></td>
</tr>
<tr>
<td>:app:createDebugCompatibleScreenManifests</td>
<td>0.011s</td>
<td></td>
</tr>
<tr>
<td>:app:dataBindingMergeGenClassesDebug</td>
<td>0.011s</td>
<td></td>
</tr>
<tr>
<td>:app:extractProguardFiles</td>
<td>0.008s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugJniLibFolders</td>
<td>0.007s</td>
<td></td>
</tr>
<tr>
<td>:app:writeDebugSigningConfigVersions</td>
<td>0.007s</td>
<td></td>
</tr>
<tr>
<td>:app:writeDebugAppMetadata</td>
<td>0.005s</td>
<td></td>
</tr>
<tr>
<td>:app:extractDeepLinksDebug</td>
<td>0.004s</td>
<td></td>
</tr>
<tr>
<td>:app:generateDebugResValues</td>
<td>0.003s</td>
<td></td>
</tr>
<tr>
<td>:app:mergeDebugNativeDebugMetadata</td>
<td>0.002s</td>
<td>NO-SOURCE</td>
</tr>
<tr>
<td>:app:compileDebugAidl</td>
<td>0.001s</td>
<td>NO-SOURCE</td>
</tr>
<tr>
<td>:app:compileDebugShaders</td>
<td>0.001s</td>
<td>NO-SOURCE</td>
</tr>
<tr>
<td>:app:compileDebugSources</td>
<td>0.001s</td>
<td>Did No Work</td>
</tr>
<tr>
<td>:app:generateDebugResources</td>
<td>0.001s</td>
<td>Did No Work</td>
</tr>
<tr>
<td>:app:preBuild</td>
<td>0.001s</td>
<td>UP-TO-DATE</td>
</tr>
<tr>
<td>:app:processDebugJavaRes</td>
<td>0.001s</td>
<td>NO-SOURCE</td>
</tr>
<tr>
<td>:app:assembleDebug</td>
<td>0s</td>
<td>Did No Work</td>
</tr>
<tr>
<td>:app:compileDebugRenderscript</td>
<td>0s</td>
<td>NO-SOURCE</td>
</tr>
<tr>
<td>:app:generateDebugAssets</td>
<td>0s</td>
<td>UP-TO-DATE</td>
</tr>
<tr>
<td>:app:preDebugBuild</td>
<td>0s</td>
<td>Did No Work</td>
</tr>
<tr>
<td>:</td>
<td>0s</td>
<td>(total)</td>
</tr>
</tbody></table>
</details>

Note that 
- `gradle assemble` assembles the code for both the `release` build config and the `debug` release config. You only need `debug` for developing. It took about 12 minutes right now (no idea where the extra ~6 minutes came from).
- `gradle build` *additionally* assembles the test code for both configs, executes the tests and runs lint.  It took about 16 minutes right now.

To find what takes so long, add `--profile` to the gradle task, e.g. `gradle assembleDebug --profile`. A HTML page with a report will be generated.

**Reply by smichel17** [reply]
What's the definition of "clean" here? Fresh clone of the repo? Deleting `build/` and `.gradle/` folders? Running `./gradlew app:clean app:cleanBuildCache`?

**Reply by westnordost** [reply]
`gradle clean`

**Comment by smichel17** [comment]
My machine: Intel i7-6700, 48 GB RAM :sweat_smile:, built on NVMe SSD

Command to build: `./gradlew :app:assembleDebug --profile`
Command to clean: `./gradlew clean`

I did a build which I didn't time, then did **three builds**, each with a clean beforehand.
For the second build, I also ran with `--no-daemon` to see if that made a difference. I got:

1. 2m30.35s
2. 1m57.44s (`--no-daemon`)
3. 1m27.64s

Sooo, clearly there's a huge speedup for subsequent builds, even if they are clean. I'd guess a lot of things are being cached in my frankly excessive amount of ram, even though the build is ostensibly clean.

edit: I can share the full details if you want, but the distribution in all cases pretty much mirrors what @westnordost posted.

**Comment by westnordost** [comment]
My current machine (AMD Ryzen 7 5700X, 32GB RAM):

- `gradle assembleDebug` after clean  requires 2m 44s
- subsequent build requires 1m 28s

What I continue to find weird is that subsequent builds to take so long. When I just change one file and recompile, I'd expect it to mostly just recompile that one file since everything else should be in byte code already.
